wxBitmap fixes for wxQt

AliKet opened the pull request:
Although the tests are all passing now, testing and reviewing this PR before merging is really appreciated, TIA!

```
> xvfb-run -a -s '-screen 0 1600x1200x24' ./test_gui [bitmap]
Test program for wxWidgets GUI features
build: 3.3.0 (UTF-8,compiler with C++ ABI compatible with gcc 4,STL containers,compatible with 3.2)
compiled using gcc 8.4
running under Linux 4.15.0-213-generic i686 as aliket
Filters: [bitmap]
===============================================================================
All tests passed (4046 assertions in 14 test cases)
```

Commit:
Micro-optimization: pre-increment instead of post-increment in loops

Commit:
For images with alpha, premultiplied ARGB32 are used now instead of ARGB32

Using premultiplied ARGB32 for image composition using alpha blending
(a more common operation) is known to be faster than using plain ARGB32

Commit:
Fix wxMask creation when converting from wxImage to wxBitmap in wxQt

wxBitmap::GetSubBitmap() should return the sub-mask too if a mask exists

Minor fix to wxMask::InitFromColour() to create a mask that conforms
to wx expectations

Commit:
Fix drawing bitmap with mask due to the previous commit

Commit:
Fix access to raw bitmap data after the switch to using premultiplied format

Commit:
Try to load/save bitmap file using Qt first

Commit:
Avoid default dithering when loading/creating monochrome bitmaps

Commit:
Enable raw bitmap data access for monochrome bitmaps in wxQt

Commit:
Fix wxBitmap conversion to wxImage with/without a mask

Commit:
Update bitmap tests for wxQt

AliKet commented on tests/graphics/bitmap.cpp:
```diff
@@ -819,7 +819,7 @@ TEST_CASE("BitmapTestCase::DrawAlpha", "[bitmap][draw][alpha]")
-#if defined(__WXMSW__) || defined(__WXOSX__)
+#if defined(__WXMSW__) || defined(__WXOSX__) || defined(__WXQT__)
     bmp.UseAlpha();
 #endif // __WXMSW__ || __WXOSX__
```
Oops! this call is not needed in `wxQt`.

AliKet commented on tests/graphics/bitmap.cpp:
```diff
@@ -934,7 +934,7 @@ TEST_CASE("BitmapTestCase::DrawAlphaWithMask", "[bitmap][draw][alpha][withmask]"
-#if defined(__WXMSW__) || defined(__WXOSX__)
+#if defined(__WXMSW__) || defined(__WXOSX__) || defined(__WXQT__)
      bmp.UseAlpha();
 #endif // __WXMSW__ || __WXOSX__
```
Not needed in wxQt

vadz commented on include/wx/rawbmp.h:
```diff
@@ -678,7 +696,7 @@ struct wxPixelDataOut<wxBitmap>
 
         // the bitmap we're associated with
-        wxBitmap m_bmp;
+        wxBitmap& m_bmp;
```
Sorry, why? I don't see anything that requires changing this bitmap now.

AliKet replied:
I thought the lack of the call to `AllocExclusive()` in `wxBitmap::GetRawData()` 
was a mistake (unless it's by design) because accessing raw bitmap data should do a **COW** to break data sharing which is not the case currently. So if we (you?) decide to change `GetRawData()` implementation for each port to follow **wxDFB**:
https://github.com/wxWidgets/wxWidgets/blob/eb8a57760e61e72959593200d4df8d8b36dfb87a/src/dfb/bitmap.cpp#L506
then this change is necessary. If not, then I can revert this if you like ?

vadz commented on include/wx/rawbmp.h:
```diff
@@ -722,7 +740,11 @@ struct wxPixelDataOut<wxBitmap>
                 operator bool() const
                 {
                     wxByte mask = static_cast<wxByte>(1 << m_bit);
+                #ifndef __WXQT__
```
This looks like the indicator that a mask in wxQt is reversed compared to wx conventions which is surely problematic and it looks like it would be better to fix it by inverting the mask when constructing it?

AliKet replied:
It's unfortunate that we have to reverse the logic here, because the code checking to see whether the current pixel is transparent or not uses this `bool operator` to determine that. and **Qt** and **wx** seem to operate in reverse manner when applying the mask  to bitmap.

vadz commented on src/qt/bitmap.cpp:
```diff
@@ -71,47 +74,63 @@ static wxImage ConvertImage( QImage qtImage )
+
+            colour = (a << 24) + (r << 16) + (g << 8) + b;
+
+            if ( !qtMask.isNull() )
+            {
+                if ( qRgb(r, g, b) != maskedColour )
+                    qtMask.setPixel(x, y, Qt::color1); // set to 1
+            }
```
I.e. maybe the logic here needs to be reversed?

vadz commented on tests/graphics/bitmap.cpp:
```diff
@@ -306,13 +306,13 @@ TEST_CASE("BitmapTestCase::ToImage", "[bitmap][image][convertto]")
                 wxColour bmpc(iBmp.Red(), iBmp.Green(), iBmp.Blue(), iBmp.Alpha());
                 wxColour imgc(image.GetRed(x, y), image.GetGreen(x, y), image.GetBlue(x, y), image.GetAlpha(x,y));
-#if defined(__WXMSW__) || defined(__WXOSX__)
+#if defined(__WXMSW__) || defined(__WXOSX__) //|| defined(__WXQT__)
```
It would be nice to leave a comment explaining why is wxQt excluded here unlike everywhere else...

vadz reviewed (commented):
Thanks for working on this! I don't know anything about Qt bitmaps/masks formats, but I find it suspicious that we must have wxQt-specific logic in common code. We can live with it if we really have to, but it would be nicer if we could avoid it.

Commit:
Fix converting wxBitmap to wxImage

After the switch to the premultiplied ARGB32 format, we should unpremultiply
the colour value when converting wxBitmap to wxImage

Also re-enable the code testing this feature and update the comments in the test code

AliKet commented:
I converted this to draft for now in the hope I can figure out how to get rid of the hack in `operator bool()`

Commit:
Fix wxGraphicsBitmap creation after recent changes to wxBitmap under wxQt

AliKet commented:
@vadz I managed to get things work without the hack in `operator bool()`. but the `graphbitmap` test fails now. I had to change this line: https://github.com/wxWidgets/wxWidgets/blob/eb8a57760e61e72959593200d4df8d8b36dfb87a/src/generic/graphicc.cpp#L1738C1-L1738C1
 to:
`                if (p.Red()+p.Green()+p.Blue() != 0)`

for the test to succeed. because the mask under wxQt is inverted, which means non-zero values in the mask (white colour) represent transparent regions in the bitmap instead of zero values. other then that everything seems to work correctly with the new changes.

My question now: is it ok to change the aforementioned line as I said above ? if yes, I'll create a new PR soon with the necessary changes only ?

vadz commented:
Yes, I think it's ok to change this, although we might want to add some helper `IsMasked()` that would be defined differently for wxQt to make this more clear.

Thanks!

AliKet commented:
Replaced bu #23872
